comms/dht/src/network_discovery/mod.rs

```diff
@@ -0,0 +1,762 @@
+                    self.publish_event(DhtEvent::NetworkDiscoveryPeersAdded(stats.clone()));
+                }
+
+                State::Ready(DiscoveryReady::new(self.context.clone(), stats))
+            },
+            (State::Ready(_), StateEvent::Idle) => State::Waiting(Waiting::new(config.idle_period)),
+            (_, StateEvent::Shutdown) => State::Shutdown,
+            (_, StateEvent::Errored(err)) => {
+                debug!(
+                    target: LOG_TARGET,
+                    "Network discovery errored: {}. Waiting for {:.0?}", err, config.on_failure_idle_period
+                );
+                State::Waiting(Waiting::new(config.on_failure_idle_period))
+            },
+            (state, event) => {
+                // TODO: impl Display
+                debug!(
+                    target: LOG_TARGET,
+                    "No state transition for event `{:?}`. The current state is `{:?}`", event, state
+                );
+                state
+            },
+        }
+    }
+
+    fn publish_event(&self, event: DhtEvent) {
+        let _ = self.event_tx.send(Arc::new(event));
+    }
+
+    #[inline]
+    fn config(&self) -> &DhtConfig {
+        &self.context.config
+    }
+
+    pub fn spawn(self) -> task::JoinHandle<()> {
+        task::spawn(self.run())
+    }
+
+    pub async fn run(mut self) {
+        if !self.config().network_discovery.enabled {
+            warn!(
+                target: LOG_TARGET,
+                "Network discovery is disabled. This node may fail to participate in the network."
+            );
+
+            return;
+        }
+        let mut state = State::Initializing;
+        loop {
+            let shutdown_signal = self.shutdown_signal.clone();
+            let next_event = {
+                let fut = self.get_next_event(&mut state);
+                futures::pin_mut!(fut);
+                or_shutdown(shutdown_signal, fut).await
+            };
+            state = self.transition(state, next_event);
+            if state.is_shutdown() {
+                break;
+            }
+        }
+    }
+}
+
+async fn or_shutdown<Fut>(shutdown_signal: ShutdownSignal, fut: Fut) -> StateEvent
+where Fut: Future<Output = StateEvent> + Unpin {
+    match future::select(shutdown_signal, fut).await {
+        Either::Left(_) => StateEvent::Shutdown,
+        Either::Right((event, _)) => event,
+    }
+}
+
+#[derive(Debug, Clone)]
+struct DiscoveryParams {
+    pub peers: Vec<NodeId>,
+    pub num_peers_to_request: usize,
+    pub max_accept_closer_peers: usize,
+}
+
+impl Display for DiscoveryParams {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "DiscoveryParams({} peer(s) selected, num_peers_to_request = {}, max_accept_closer_peers = {})",
+            self.peers.len(),
+            self.num_peers_to_request,
+            self.max_accept_closer_peers
+        )
+    }
+}
+
+#[derive(Debug)]
+struct Initializing<'a> {
```

Comment:
It would be helpful to have these in different files to help future source control diffs

Reply:
This was bothering me too :) Will make the change